src/theory/arith/nl/cad/cdcac_utils.cpp

```diff
@@ -0,0 +1,267 @@
+  }
+  if (c > 0)
+  {
+    Trace("libpoly::interval_connect")
+        << lhs << " and " << rhs << " overlap." << std::endl;
+    return true;
+  }
+  Assert(c == 0);
+  if (lhs.get_internal()->is_point || rhs.get_internal()->is_point
+      || !lhs.get_internal()->b_open || !rhs.get_internal()->a_open)
+  {
+    Trace("libpoly::interval_connect")
+        << lhs << " and " << rhs
+        << " touch and the intermediate point is covered." << std::endl;
+    return true;
+  }
+  return false;
+}
+
+void clean_intervals(std::vector<CACInterval>& intervals)
+{
+  // Simplifies removal of redundancies later on.
+  if (intervals.size() < 2) return;
+
+  // Sort intervals.
+  std::sort(intervals.begin(),
+            intervals.end(),
+            [](const CACInterval& lhs, const CACInterval& rhs) {
+              return compare_for_cleanup(lhs.mInterval, rhs.mInterval);
+            });
+
+  // Remove intervals that are covered by others.
+  // Implementation roughly follows
+  // https://en.cppreference.com/w/cpp/algorithm/remove Find first interval that
+  // covers the next one.
+  std::size_t first = 0;
+  for (; first < intervals.size() - 1; ++first)
+  {
+    if (interval_covers(intervals[first].mInterval,
+                        intervals[first + 1].mInterval))
+    {
+      break;
+    }
+  }
+  // If such an interval exists, remove accordingly.
+  if (first < intervals.size() - 1)
+  {
+    for (std::size_t i = first + 2; i < intervals.size(); ++i)
+    {
+      if (!interval_covers(intervals[first].mInterval, intervals[i].mInterval))
+      {
+        // Interval is not covered. Move it to the front and bump front.
+        ++first;
+        intervals[first] = std::move(intervals[i]);
+      }
+      // Else: Interval is covered as well.
+    }
+    // Erase trailing values
+    while (intervals.size() > first + 1)
+    {
+      intervals.pop_back();
+    }
+  }
+}
+
+std::vector<Node> collect_constraints(const std::vector<CACInterval>& intervals)
+{
+  std::vector<Node> res;
+  for (const auto& i : intervals)
+  {
+    res.insert(res.end(), i.mOrigins.begin(), i.mOrigins.end());
+  }
+  std::sort(res.begin(), res.end());
+  auto it = std::unique(res.begin(), res.end());
+  res.erase(it, res.end());
+  return res;
+}
+
+bool sample_outside(const std::vector<CACInterval>& infeasible, Value& sample)
+{
+  if (infeasible.empty())
+  {
+    // No infeasible region, just take anything: zero
+    sample = poly::Integer();
+    return true;
+  }
+  if (!is_minus_infinity(get_lower(infeasible.front().mInterval)))
+  {
+    // First does not cover -oo, just take sufficiently low value
+    Trace("cdcac") << "Sample before " << infeasible.front().mInterval
+                   << std::endl;
+    const auto* i = infeasible.front().mInterval.get_internal();
+    sample = value_between(
+        Value::minus_infty().get_internal(), true, &i->a, !i->a_open);
+    return true;
+  }
+  for (std::size_t i = 0; i < infeasible.size() - 1; ++i)
```

Comment:
Cache size in separate variable.
```suggestion
  for (std::size_t i = 0, size =  infeasible.size(); i < size - 1; ++i)
```